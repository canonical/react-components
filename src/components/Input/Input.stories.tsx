import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Input from "./Input";

const meta: Meta<typeof Input> = {
  component: Input,
  tags: ["autodocs"],

  argTypes: {
    caution: {
      control: {
        type: "text",
      },
    },

    error: {
      control: {
        type: "text",
      },
    },

    help: {
      control: {
        type: "text",
      },
    },

    helpClassName: {
      control: {
        type: "text",
      },
    },

    label: {
      control: {
        type: "text",
      },
    },

    success: {
      control: {
        type: "text",
      },
    },

    id: {
      control: {
        disable: true,
      },
    },

    placeholder: {
      control: {
        type: "text",
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Input>;

export const TextInput: Story = {
  name: "Text input",

  args: {
    type: "text",
    id: "exampleTextInput2",
    placeholder: "example@canonical.com",
    label: "Email address",
    help: "Additional description for the field",
    helpClassName: "u-no-margin--bottom",
  },
};

export const Stacked: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput3"
      placeholder="example@canonical.com"
      stacked
      label="Email address"
    />
  ),

  name: "Stacked",
};

export const Disabled: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput4"
      placeholder="example@canonical.com"
      disabled
      label="Email address"
    />
  ),

  name: "Disabled",
};

export const Error: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput5"
      placeholder="example@canonical.com"
      error="This field is required."
      label="Email address"
    />
  ),

  name: "Error",
};

export const Success: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput6"
      placeholder="example@canonical.com"
      success="Verified."
      label="Email address"
    />
  ),

  name: "Success",
};

export const Caution: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput7"
      placeholder="example@canonical.com"
      caution="No validation is performed in preview mode."
      label="Email address"
    />
  ),

  name: "Caution",
};

export const Required: Story = {
  render: () => (
    <Input
      type="text"
      id="exampleTextInput8"
      placeholder="example@canonical.com"
      required
      label="Email address"
    />
  ),

  name: "Required",
};

export const Checkbox: Story = {
  render: () => (
    <Input type="checkbox" id="checkExample12" defaultChecked label="HTML" />
  ),
  name: "Checkbox",
};

export const RadioButton: Story = {
  render: () => (
    <Input
      label="Linux"
      type="radio"
      name="RadioOptions"
      id="Radio12"
      defaultValue="option1"
      defaultChecked
      help="Ubuntu"
    />
  ),

  name: "Radio button",
};
