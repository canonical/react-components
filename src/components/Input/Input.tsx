import classNames from "classnames";
import React, { useEffect, useRef } from "react";
import type { InputHTMLAttributes, ReactNode } from "react";

import Field from "../Field";
import CheckboxInput from "../CheckboxInput";
import RadioInput from "../RadioInput";

import type { ClassName, PropsWithSpread } from "types";
import { useId } from "hooks";

/**
 * The props for the Input component.
 */
export type Props = PropsWithSpread<
  {
    /**
     * The content for caution validation.
     */
    caution?: ReactNode;
    /**
     * Optional class(es) to pass to the input element.
     */
    className?: ClassName;
    /**
     * The content for error validation message. Controls the value of aria-invalid attribute.
     */
    error?: ReactNode;
    /**
     * Help text to show below the field.
     */
    help?: ReactNode;
    /**
     * Optional class(es) to pass to the help text element.
     */
    helpClassName?: ReactNode;
    /**
     * The id of the input.
     */
    id?: string;
    /**
     * The label for the field.
     */
    label?: ReactNode;
    /**
     * Optional class(es) to pass to the label component.
     */
    labelClassName?: string;
    /**
     * Whether the field is required.
     */
    required?: boolean;
    /**
     * Whether the form field should have a stacked appearance.
     */
    stacked?: boolean;
    /**
     * The content for success validation.
     */
    success?: ReactNode;
    /**
     * Whether to focus on the input on initial render.
     */
    takeFocus?: boolean;
    /**
     * Optional class(es) to pass to the wrapping Field component
     */
    wrapperClassName?: string;
  },
  InputHTMLAttributes<HTMLInputElement>
>;

const Input = ({
  caution,
  className,
  error,
  help,
  helpClassName,
  id,
  label,
  labelClassName,
  required,
  stacked,
  success,
  takeFocus,
  type,
  wrapperClassName,
  ...inputProps
}: Props): JSX.Element => {
  const inputRef = useRef(null);
  const fieldLabel = !["checkbox", "radio"].includes(type) ? label : "";
  const validationId = useId();
  const helpId = useId();
  const hasError = !!error;

  const commonProps = {
    "aria-describedby": [help ? helpId : null, success ? validationId : null]
      .filter(Boolean)
      .join(" "),
    "aria-errormessage": hasError ? validationId : null,
    "aria-invalid": hasError,
    id: id,
    label: label,
    required: required,
    ...inputProps,
  };

  useEffect(() => {
    if (takeFocus) {
      inputRef.current.focus();
    }
  }, [takeFocus]);

  let input: ReactNode;
  if (type === "checkbox") {
    input = (
      <CheckboxInput
        label={label}
        labelClassName={labelClassName}
        {...commonProps}
      />
    );
  } else if (type === "radio") {
    input = (
      <RadioInput
        label={label}
        labelClassName={labelClassName}
        {...commonProps}
      />
    );
  } else {
    input = (
      <input
        className={classNames("p-form-validation__input", className)}
        ref={inputRef}
        type={type}
        {...commonProps}
      />
    );
  }
  return (
    <Field
      caution={caution}
      className={wrapperClassName}
      error={error}
      forId={id}
      help={help}
      helpClassName={helpClassName}
      helpId={helpId}
      isTickElement={type === "checkbox" || type === "radio"}
      label={fieldLabel}
      labelClassName={labelClassName}
      required={required}
      stacked={stacked}
      success={success}
      validationId={validationId}
    >
      {input}
    </Field>
  );
};

export default Input;
