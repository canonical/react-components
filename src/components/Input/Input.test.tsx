import { mount, shallow } from "enzyme";
import React from "react";

import Input from "./Input";

describe("Input ", () => {
  it("renders", () => {
    const wrapper = shallow(<Input type="text" id="test-id" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    const wrapper = shallow(<Input type="text" className="extra-class" />);
    const className = wrapper.find("input").prop("className");
    expect(className.includes("p-form-validation__input")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });

  it("moves the label for radio buttons", () => {
    const wrapper = shallow(<Input type="radio" />);
    expect(wrapper.prop("labelFirst")).toBe(false);
  });

  it("moves the label for checkboxes", () => {
    const wrapper = shallow(<Input type="checkbox" />);
    expect(wrapper.prop("labelFirst")).toBe(false);
  });

  it("can take focus on first render", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    const wrapper = mount(<Input takeFocus />, { attachTo: container });
    expect(wrapper.find("input").getDOMNode()).toBe(document.activeElement);
  });

  it("sets aria-invalid to false when there is no error", () => {
    const wrapper = mount(<Input type="text" />);
    expect(wrapper.find("input").prop("aria-invalid")).toBe(false);
  });

  it("sets aria-invalid to true when there is an error", () => {
    const wrapper = mount(<Input type="text" error="Incorrect value" />);
    expect(wrapper.find("input").prop("aria-invalid")).toBe(true);
  });
});
