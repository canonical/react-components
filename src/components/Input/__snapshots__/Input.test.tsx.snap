// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input RTL renders 1`] = `
<div>
  <div
    class="p-form__group p-form-validation"
  >
    <div
      class="p-form__control u-clearfix"
    >
      <input
        aria-describedby=""
        aria-invalid="false"
        class="p-form-validation__input"
        id="test-id"
        type="text"
      />
    </div>
  </div>
</div>
`;
